### DISCUSSION TITLE
Harvest Power Dashboard

### DISCUSSION THREAD

**Discussion by cgrinds** [NetApp/harvest#951] [Announcements]
The Harvest team has added a new environmental dashboard to Harvest targeted for the May 2022 release.  _If you want to try this dashboard out before the next release, you can grab the latest [nightly build](https://github.com/NetApp/harvest/releases/tag/nightly)._

Below you can see an example of the dashboard with several of its highlighted features:
- Total power usage (node + shelf)
- Shelf and node temperature and fan speeds
- Table details on cluster/node/shelf power, temp, fan speeds
- Similar power metrics are also included on the Cluster and Shelf dashboards
 
We'd love to hear what you think.
# Screenshots
![image](https://user-images.githubusercontent.com/242252/161816171-23a2ff30-43ce-413a-85f0-fbce277879ef.png)

![image](https://user-images.githubusercontent.com/242252/161816203-8d66787e-b760-416a-b886-644ad48ee9e3.png)

![image](https://user-images.githubusercontent.com/242252/161823532-61fb7806-ee8b-485d-934c-e75ba6875bdf.png)

**Comment by cgrinds** [comment]
Some hardware models don't return data for power, voltage, or current metrics. When that happens, Harvest is not able to calculate the power and shows zeroes in the data. We're working on improvements here.

**Comment by buddlflink** [comment]
Is there a way for a more finegrained view, to break down power consumptions to SVM level or even volumes, e. g. based on capacities and CPU load?

**Comment by cgrinds** [comment]
hi @buddlflink good question - we'll take a look. If you get a chance, please open a feature request with that so we can track.

**Comment by ewilts** [comment]
I have 2 AFF800 clusters.  It's reporting 0 power for the HA pairs but is reporting power for the few attached shelves I have.  Should the nodes be reporting power?

**Reply by cgrinds** [reply]
hi @ewilts, yes, the nodes should be reporting power for AFF800 clusters. As @rahulguptajss mentioned, if you can share the poller logs we can take a look.

**Comment by rahulguptajss** [comment]
@ewilts Could you share harvest logs for these clusters.

**Reply by ewilts** [reply]
nabox-harvest2        | 8:31PM WRN goharvest2/cmd/collectors/zapi/plugins/shelf/shelf.go:369 > no [cooling-fans] instances on this system Poller=blue object=Shelf plugin=Zapi:Shelf
nabox-harvest2        | 8:31PM WRN goharvest2/cmd/collectors/zapi/plugins/shelf/shelf.go:369 > no [current-sensors] instances on this system Poller=blue object=Shelf plugin=Zapi:Shelf
nabox-harvest2        | 8:31PM WRN goharvest2/cmd/collectors/zapi/plugins/shelf/shelf.go:369 > no [power-supply-units] instances on this system Poller=blue object=Shelf plugin=Zapi:Shelf
nabox-harvest2        | 8:31PM WRN goharvest2/cmd/collectors/zapi/plugins/shelf/shelf.go:369 > no [temperature-sensors] instances on this system Poller=blue object=Shelf plugin=Zapi:Shelf
nabox-harvest2        | 8:31PM WRN goharvest2/cmd/collectors/zapi/plugins/shelf/shelf.go:369 > no [voltage-sensors] instances on this system Poller=blue object=Shelf plugin=Zapi:Shelf

Are there other specific messages I need to be looking for?  This is ONTAP 9.9.1P6 BTW.

**Reply by ewilts** [reply]
I also have lot of these (for every sensor):
nabox-harvest2        | 9:57PM ERR goharvest2/cmd/collectors/zapi/collector/zapi.go:425 > skipped instance [blue-01.Bat Charge FET]: not found in cache Poller=blue collector=Zapi:Sensor
nabox-harvest2        | 9:57PM ERR goharvest2/cmd/collectors/zapi/collector/zapi.go:425 > skipped instance [blue-01.Bat Cycle Count]: not found in cache Poller=blue collector=Zapi:Sensor
nabox-harvest2        | 9:57PM ERR goharvest2/cmd/collectors/zapi/collector/zapi.go:425 > skipped instance [blue-01.Attention]: not found in cache Poller=blue collector=Zapi:Sensor

**Comment by ewilts** [comment]
I think I fixed it - it was a permission error in the roles.  When I updated my role, it started working.
New role:
security login role create -role netapp-harvest2 -access readonly -cmddirname DEFAULT

**Reply by rahulguptajss** [reply]
> I think I fixed it - it was a permission error in the roles. When I updated my role, it started working. New role: security login role create -role netapp-harvest2 -access readonly -cmddirname DEFAULT

Great! Ideally subset of permissions mentioned [here](https://github.com/NetApp/harvest/blob/main/docs/AuthAndPermissions.md) should also be good enough instead of the DEFAULT permissions.

**Comment by mjgalloway** [comment]
@cgrinds - What API or sensor data are you collecting to calculate the power? My customer has a home-grown app the wrote and would like to power metrics for NetApp.

**Reply by cgrinds** [reply]
hi @mjgalloway Take a look at https://netapp.github.io/harvest/nightly/resources/power-algorithm/ and see if that answers you questions.

**Reply by mjgalloway** [reply]
Thank you. I'll pass this along and see if this is what they are looking for, just from a quick look I would say it does.

**Comment by mjgalloway** [comment]
@cgrinds - My customer ran the API and collected the below and are now asking if they need to do any filtering before applying the algorithm.

[swdctinasf21_power_draw.xlsx](https://github.com/NetApp/harvest/files/12797227/swdctinasf21_power_draw.xlsx)

**Reply by cgrinds** [reply]
hi @mjgalloway there are several gotchas outlined on https://netapp.github.io/harvest/nightly/resources/power-algorithm/ that will need to be taken into account. Can they use Harvest for this metric instead of re-implementing?

The logic Harvest uses to calculate power is found in:
- https://github.com/NetApp/harvest/blob/main/cmd/collectors/power.go - used to calculate node power
- https://github.com/NetApp/harvest/blob/main/cmd/collectors/restperf/plugins/disk/disk.go - used to calculate disk shelf power

**Reply by mjgalloway** [reply]
Thank you, Chris. They have a homegrown app that that want to use. I'll pass this info along and see if they make any headway. I suspect I'll have to open a resource request since they have already asked about speaking to engineering on how to implement this.

**Comment by Harriwaltan** [comment]
I have a question concerning a power dashboard for StorageGrid. Is there any plan to implement the necessary metrices for this - comparable to NetApp OnTap? Is there a roadmap for this?
Thanks for your feedback.

**Reply by cgrinds** [reply]
hi @Harriwaltan that's a great idea, but I don't think StorageGRID provides enough metrics for those. I only see `node_hwmon_power_average_watt` and `node_hwmon_temp_celsius`. Are you aware of others?

**Comment by Harriwaltan** [comment]
I spoke with my presales engineer about this topic. He confirmed that the relevant metrics are missing for StorageGrid. He will open a future enhancement request to StorageGird engineering.

**Reply by cgrinds** [reply]
Sounds good
